crates/sui-core/src/authority.rs

```diff
@@ -2079,6 +2084,23 @@ impl AuthorityState {
         Ok(())
     }
 
+    /// Returns true if certificate is a shared-object cert but has not been sequenced.
+    async fn is_cert_awaiting_sequencing(
```

Comment:
The implementation of this function is slightly different from the part in prepare_certificate, could you double check that they are consistent (and perhaps reuse code if possible)?

Reply:
They are consistent as far as I can tell - the main difference is that we don't want to actually try to load the shared locks pre-verification (since an unverified cert could bypass any limits on input counts that we later add).

As a TODO, if this function returns true, it is guaranteed that the locks exist, so we could remove several error checking paths and turn them into asserts.

Reply:
This is probably an edge case, but I wonder if we somehow got the epoch change tx through gossip instead of normal local execution, we shouldn't need to wait for sequencing.

Reply:
@lxfind added a commit for this, please take a look